Mathlib/Data/Multiset/Basic.lean

```diff
@@ -2316,6 +1998,297 @@ theorem filter_eq' (s : Multiset α) (b : α) : s.filter (· = b) = replicate (c
 theorem filter_eq (s : Multiset α) (b : α) : s.filter (Eq b) = replicate (count b s) b := by
   simp_rw [← filter_eq', eq_comm]
 
+lemma erase_attach_map_val (s : Multiset α) (x : {x // x ∈ s}) :
+    (s.attach.erase x).map (↑) = s.erase x := by
+  rw [Multiset.map_erase _ val_injective, attach_map_val]
+
+lemma erase_attach_map (s : Multiset α) (f : α → β) (x : {x // x ∈ s}) :
+    (s.attach.erase x).map (fun j : {x // x ∈ s} ↦ f j) = (s.erase x).map f := by
+  simp only [← Function.comp_apply (f := f)]
+  rw [← map_map, erase_attach_map_val]
+
+end
+
+/-! ### Subtraction -/
+
+section sub
+variable [DecidableEq α] {s t u : Multiset α} {a : α}
+
+/-- `s - t` is the multiset such that `count a (s - t) = count a s - count a t` for all `a`
+  (note that it is truncated subtraction, so it is `0` if `count a t ≥ count a s`). -/
+protected def sub (s t : Multiset α) : Multiset α :=
+  (Quotient.liftOn₂ s t fun l₁ l₂ => (l₁.diff l₂ : Multiset α)) fun _v₁ _v₂ _w₁ _w₂ p₁ p₂ =>
+    Quot.sound <| p₁.diff p₂
+
+instance : Sub (Multiset α) :=
+  ⟨Multiset.sub⟩
+
+@[simp]
+lemma coe_sub (s t : List α) : (s - t : Multiset α) = (s.diff t : List α) :=
+  rfl
+
+/-- This is a special case of `tsub_zero`, which should be used instead of this.
+  This is needed to prove `OrderedSub (Multiset α)`. -/
+@[simp, nolint simpNF] -- We want to use this lemma earlier than the lemma simp can prove it with
+protected lemma sub_zero (s : Multiset α) : s - 0 = s :=
+  Quot.inductionOn s fun _l => rfl
+
+@[simp]
+lemma sub_cons (a : α) (s t : Multiset α) : s - a ::ₘ t = s.erase a - t :=
+  Quotient.inductionOn₂ s t fun _l₁ _l₂ => congr_arg _ <| diff_cons _ _ _
+
+protected lemma zero_sub (t : Multiset α) : 0 - t = 0 :=
+  Multiset.induction_on t rfl fun a s ih => by simp [ih]
+
+@[simp]
+lemma countP_sub {s t : Multiset α} :
+    t ≤ s → ∀ (p : α → Prop) [DecidablePred p], countP p (s - t) = countP p s - countP p t :=
+  Quotient.inductionOn₂ s t fun _l₁ _l₂ hl _ _ ↦ List.countP_diff hl _
```

Comment:
Some more to add to the PR description: `countP_sub` now has a different argument order. (Presumably to match `List.countP_diff`?)